src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -7072,45 +7073,54 @@ std::pair<MergeTreeData::MutableDataPartPtr, scope_guard> MergeTreeData::cloneAn
     const WriteSettings & write_settings)
 {
     chassert(!isStaticStorage());
-
-    /// Check that the storage policy contains the disk where the src_part is located.
-    bool does_storage_policy_allow_same_disk = false;
-    for (const DiskPtr & disk : getStoragePolicy()->getDisks())
+    bool on_same_disk = false;
+    for (const DiskPtr & disk : this->getStoragePolicy()->getDisks())
     {
         if (disk->getName() == src_part->getDataPartStorage().getDiskName())
         {
-            does_storage_policy_allow_same_disk = true;
+            on_same_disk = true;
             break;
         }
     }
-    if (!does_storage_policy_allow_same_disk)
-        throw Exception(
-            ErrorCodes::BAD_ARGUMENTS,
-            "Could not clone and load part {} because disk does not belong to storage policy",
-            quoteString(src_part->getDataPartStorage().getFullPath()));
+
 
     String dst_part_name = src_part->getNewName(dst_part_info);
     String tmp_dst_part_name = tmp_part_prefix + dst_part_name;
     auto temporary_directory_lock = getTemporaryPartDirectoryHolder(tmp_dst_part_name);
 
     /// Why it is needed if we only hardlink files?
+    /// Answer: In issue #59377, add copy when attach from different disk.
     auto reservation = src_part->getDataPartStorage().reserve(src_part->getBytesOnDisk());
     auto src_part_storage = src_part->getDataPartStoragePtr();
 
     scope_guard src_flushed_tmp_dir_lock;
     MergeTreeData::MutableDataPartPtr src_flushed_tmp_part;
 
     String with_copy;
-    if (params.copy_instead_of_hardlink)
+    if (params.copy_instead_of_hardlink || !on_same_disk)
         with_copy = " (copying data)";
 
-    auto dst_part_storage = src_part_storage->freeze(
-        relative_data_path,
-        tmp_dst_part_name,
-        read_settings,
-        write_settings,
-        /* save_metadata_callback= */ {},
-        params);
+
+    std::shared_ptr<IDataPartStorage> dst_part_storage{};
+    if (on_same_disk && !params.copy_instead_of_hardlink)
+    {
+        dst_part_storage = src_part_storage->freeze(
+            relative_data_path,
+            tmp_dst_part_name,
+            read_settings,
+            write_settings,
+            /* save_metadata_callback= */ {},
+            params);
+    }
+    else
+    {
+        auto reservation_on_dst = getStoragePolicy()->reserve(src_part->getBytesOnDisk());
+        if (!reservation_on_dst)
+            throw Exception(ErrorCodes::NOT_ENOUGH_SPACE, "Not enough space on disk.");
+        dst_part_storage = src_part_storage->clonePart(
```

Comment:
Could you please give me more information about your thinking? 

I think that the clonePart function is used to "Make a full copy of a data part into 'to/dir_path' (possibly to a different disk)." The purpose of this PR is to attach partitions from the table on other disks using copy instead of hard link, which can use the clonePart function. Could you give me some suggestions about how to change the clonePart function if you think it is necessary?

<img width="639" alt="image" src="https://github.com/ClickHouse/ClickHouse/assets/54164389/686d1cd5-c2d1-4bf1-9c8f-5db17344c8f4">

Reply:
It seems like the freeze function has local backup logic on the same disk. Maybe we should build a freezeRemote function to retain the function points of freeze and allow backup to be performed on different disks.
<img width="603" alt="image" src="https://github.com/ClickHouse/ClickHouse/assets/54164389/e22259de-f891-49f8-890b-a3def318aa5a">

Reply:
Please check https://github.com/ClickHouse/ClickHouse/pull/63087 about freezeRemote() part. Let me know if the codes satisfy your ideas. Thank you for your time.